Add K3s Hardening Guide and Self Assessment (#2882)
rancher/docs: content/k3s/latest/en/security/hardening_guide/_index.md

---
title: "CIS Hardening Guide"
weight: 80
---

# CIS Hardening Guide

This document provides prescriptive guidance for hardening a production installation of K3s. It outlines the configurations and controls required to address Kubernetes benchmark controls from the Center for Information Security (CIS).

K3s has a number of security mitigations applied and turned on by default and will pass a number of the Kubernetes CIS controls without modification. There are some notable exceptions to this that require manual intervention to fully comply with the CIS Benchmark:

> 1. K3s will not modify the host operating system. Any host-level modifications will need to be done manually.

Reviewer: wording nit: Do you think it is accurate to say that we dont modify the host operating system? I guess it depends on definition of "host operating system," because we do lay down a bunch of binaries on the path, setup a systemd unit, etc.

If you were reading this as an operator, would that confuse you?

If so, then maybe we just say something more specific like "K3s will not modify host kernel parameters"?

I know rke2 has the same wording, so if we change it here, can you change it there too?

To be clear: I don't know the answer here, so I'm more just asking your opinion and will go with what you think works.

Author: From a Linux perspective, when I hear "operating system", I think kernel and not userland or fielsystem. But that's just me and I don't know if there's any consensus around that. Happy to get additional feedback here from others as well.

Reviewer: yeah, ask @Oats87 about the woriding explicitly. go with what he says

> **Note:** `protect-kernel-defaults` is exposed as a top-level flag for K3s.

#### Set kernel parameters

Create a file called `/etc/sysctl.d/90-kubelet.conf` and add the snippet below. Then run `sysctl -p /etc/sysctl.d/90-kubelet.conf` to 

```bash
vm.panic_on_oom=0
vm.overcommit_memory=1
kernel.panic=10
kernel.panic_on_oops=1
```

## Kubernetes Runtime Requirements

The runtime requirements to comply with the CIS Benchmark are centered around pod security (PSPs) and network policies. These are outlined in this section. K3s doesn't apply any default PSPs or network policies however K3s ships with a controller that is meant to apply a given set of network policies. By default, K3s runs with the "NodeRestriction" admission controller. To enable PSPs, add the following to the K3s start command: `--kube-apiserver-arg="enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount"`. This will have the effect of maintaining the "NodeRestriction" plugin as well as enabling the "PodSecurityPolicy".

### PodSecurityPolicies

> When PSPs are enabled, a policy can be applied to mitigate the necessary controls described in section 5.2 of the CIS Benchmark.

Reviewer: can we say satisfy instead of mitigate here?

Here's an example of a compliant PSP.

```yaml
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: cis1.5-compliant-psp
spec:
  privileged: false                # CIS - 5.2.1
  allowPrivilegeEscalation: false  # CIS - 5.2.5
  requiredDropCapabilities:        # CIS - 5.2.7/8/9
    - ALL
  volumes:
    - 'configMap'
    - 'emptyDir'
    - 'projected'
    - 'secret'
    - 'downwardAPI'
    - 'persistentVolumeClaim'
  hostNetwork: false               # CIS - 5.2.4
  hostIPC: false                   # CIS - 5.2.3
  hostPID: false                   # CIS - 5.2.2
  runAsUser:
    rule: 'MustRunAsNonRoot'       # CIS - 5.2.6
  seLinux:
    rule: 'RunAsAny'
  supplementalGroups:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  fsGroup:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  readOnlyRootFilesystem: false
```

> **Note:** The Kubernetes critical additions such as CNI, DNS, and Ingress are ran as pods in the `kube-system` namespace. Therefore, this namespace will have a policy that is less restrictive so that these components can run properly.

### NetworkPolicies

> CIS requires that all namespaces have a network policy applied that reasonably limits traffic into namespaces and pods.

Reviewer: ```suggestion
> NOTE: K3s's deploys kube-router for network policy enforcement. Support for this in K3s is currently experimental.

CIS requires that all namespaces have a network policy applied that reasonably limits traffic into namespaces and pods.
```

I'm not sure that the above note will render correctly or if it follows existing patterns for "notes" in our docs. Can you make sure of that before you add it in?

Author: This appears to be convention in this repo throughout the Rancher v2.x markdown files.

Here's an example of a compliant network policy. 

```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: intra-namespace
  namespace: kube-system
spec:
  podSelector:
  ingress:
    - from:
      - NamespaceSelector:
          matchLabels:
            name: kube-system
```

> **Note:** Operators must manage network policies as normal for additional namespaces that are created.

## Known Issues
The following are controls that K3s currently does not pass by default. Each gap will be explained, along with a note clarifying whether it can be passed through manual operator intervention, or if it will be addressed in a future release of K3s.

### Control 1.2.15

> Ensure that the admission control plugin `NamespaceLifecycle` is set.

Reviewer: This is enabled by default in kubernetes. are we doing something to disable it? Does setting our own (to enable PSP) cause the defaults to be turned off?

---
title: "CIS Hardening Guide"
weight: 80
---

# CIS Hardening Guide

This document provides prescriptive guidance for hardening a production installation of K3s. It outlines the configurations and controls required to address Kubernetes benchmark controls from the Center for Information Security (CIS).

K3s has a number of security mitigations applied and turned on by default and will pass a number of the Kubernetes CIS controls without modification. There are some notable exceptions to this that require manual intervention to fully comply with the CIS Benchmark:

1. K3s will not modify the host operating system. Any host-level modifications will need to be done manually.

> 2. Certain CIS policy controls for PodSecurityPolicies and NetworkPolicies will restrict the functionality of this cluster. You must opt into having K3s configure these by adding the appropriate options to your command-line flags or configuration file.

Reviewer: What command line flags are you referring to here? Possible this is a hold-over from rke2 docs and it doesnt apply to k3s?

Author: The CLI flags in question here are in reference to the additional flags that can be passed to the control plane components. There's an example at the bottom for reference.

Reviewer: Which flags specifically lets k3s configure PSPs or network policies?

I dont think k3s configures either of thsese.

Author: None directly. I'll reword it. It was in reference to this bit specifically which would enable their use. `--kube-apiserver-arg="enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount"` Admittedly, not great wording.

> 1. K3s will not modify the host operating system. Therefore, you, the operator, must make a few Host-level modifications.
> 2. Certain CIS policy controls for PodSecurityPolicies and NetworkPolicies will restrict the functionality of this cluster. You must opt into having K3s configuring these.
> 
> The first section of the CIS Benchmark (1.1), concerns itself primarily with pod manifest permissions and ownership. K3s doesn't utilize these for the core components since everything is packaged into a single binary.

Reviewer 2: This scans a bit weird to me, how about:

```suggestion
1. K3s will not modify the host operating system. Any host-level modifications will need to be done manually.
2. Certain CIS policy controls for PodSecurityPolicies and NetworkPolicies will restrict the functionality of this cluster. You must opt into having K3s configure these by adding the appropriate options to your command-line flags or configuration file.

The first section (1.1) of the CIS Benchmark concerns itself primarily with pod manifest permissions and ownership. K3s doesn't utilize these for the core components since everything is packaged into a single binary.
```

This is a kubelet flag that will cause the kubelet to exit if the required kernel parameters are unset or are set to values that are different from the kubelet's defaults.

> **Note:** `protect-kernel-defaults` is exposed as a top-level flag for K3s.

#### Set kernel parameters

Create a file called `/etc/sysctl.d/90-kubelet.conf` and add the snippet below. Then run `sysctl -p /etc/sysctl.d/90-kubelet.conf` to 

```bash
vm.panic_on_oom=0
vm.overcommit_memory=1
kernel.panic=10
kernel.panic_on_oops=1
```

## Kubernetes Runtime Requirements

The runtime requirements to comply with the CIS Benchmark are centered around pod security (PSPs)and network policies. These are outlined in this section. K3s doesn't apply any default PSPs or network policies however K3s ships with a controller that is meant to apply a given set of network policies. By default, K3s runs with the "NodeRestriction" admission controller. To enable PSPs, add the following to the K3s start command: `--kube-apiserver-arg="enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount"`. This will have the effect of maintaining the "NodeRestriction" plugin as well as enabling the "PodSecurityPolicy".

> ### PodSecurityPolicies

Reviewer: This section is inaccurate because k3s does not support the profile flag.

Author: Updated.

K3s always runs with the PodSecurityPolicy admission controller turned on. However, when it is **not** started with the cis-1.5 profile, K3s will put an unrestricted policy in place that allows Kubernetes to run as though the PodSecurityPolicy admission controller was not enabled.

When ran with the cis-1.5 profile, K3s will put a much more restrictive set of policies in place. These policies meet the requirements outlined in section 5.2 of the CIS Benchmark.

Here's an example of a compliant PSP.

```yaml
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: cis1.5-compliant-psp
  annotations:
    seccomp.security.alpha.kubernetes.io/allowedProfileNames: 'docker/default,runtime/default'

>     apparmor.security.beta.kubernetes.io/allowedProfileNames: 'runtime/default'

Reviewer 3: Are these annotations necessary for this PSP, given we know that they can cause problems on non-AppArmor enabled hosts.

Reviewer 2: I am +1 on taking these out and having them align with the PSPs that we are using for RKE2

Author: Yeah, I'll remove. This was made prior to the revert in RKE2.

Reviewer: Still need to do this?

Author: No. Removed in commit `6d47b76`.

spec:
  privileged: false                # CIS - 5.2.1
  allowPrivilegeEscalation: false  # CIS - 5.2.5
  requiredDropCapabilities:        # CIS - 5.2.7/8/9
    - ALL
  volumes:
    - 'configMap'
    - 'emptyDir'
    - 'projected'
    - 'secret'
    - 'downwardAPI'
    - 'persistentVolumeClaim'
  hostNetwork: false               # CIS - 5.2.4
  hostIPC: false                   # CIS - 5.2.3
  hostPID: false                   # CIS - 5.2.2
  runAsUser:
    rule: 'MustRunAsNonRoot'       # CIS - 5.2.6
  seLinux:
    rule: 'RunAsAny'
  supplementalGroups:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  fsGroup:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  readOnlyRootFilesystem: false
```

> **Note:** The Kubernetes control plane components and critical additions such as CNI, DNS, and Ingress are ran as pods in the `kube-system` namespace. Therefore, this namespace will have a policy that is less restrictive so that these components can run properly.

### NetworkPolicies

CIS requires that all namespaces have a network policy applied that reasonably limits traffic into namespaces and pods.

> Here's an example of a compliant network policy. 

Reviewer 2: Can we provide some information on what this PSP does?

Author: The comments in there are meant to serve that sine the accompanying doc (self assessment) has further detail.

```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: intra-namespace
  namespace: kube-system
spec:
  podSelector:
  ingress:
    - from:
      - NamespaceSelector:
          matchLabels:
            name: kube-system
```

> > **Note:** Operators must manage network policies as normal for additional namespaces that are created.

Reviewer 2: What do we mean by 'as normal'? Are we trying to say that, if they use the PSP above, they need to provide additional PSPs for all other namespaces?

## Known Issues

> The following are controls that K3s currently does not pass, by default. Each gap will be explained and whether it can be passed through manual operator intervention or if it will be addressed in a future release.

Reviewer 2: ```suggestion
The following are controls that K3s currently does not pass by default. Each gap will be explained, along with a note clarifying whether it can be passed through manual operator intervention, or if it will be addressed in a future release of K3s.
```

The first section of the CIS Bencmark, 1.1, concerns itself primarily with pod manifest permissions and ownership. K3s doesn't utilize these for the core components since everything is packaged into a single binary.

## Host-level Requirements

There are two areas of of host-level requirements: kernel parameters and etcd process/directory configuration. These are outlined in this section.

### Ensure `protect-kernel-defaults` is set

This is a kubelet flag that will cause the kubelet to exit if the required kernel parameters are unset or are set to values that are different from the kubelet's defaults.

> **Note:** `protect-kernel-defaults` is exposed as a top-level flag for K3s.

#### Set kernel parameters

Create a file called `/etc/sysctl.d/90-kubelet.conf` and add the snippet below. Then run `sysctl -p /etc/sysctl.d/90-kubelet.conf` to 

```bash
vm.panic_on_oom=0
vm.overcommit_memory=1
kernel.panic=10
kernel.panic_on_oops=1
```

### Ensure etcd is started properly

The CIS Benchmark requires that the etcd data directory be owned by the `etcd` user and group.

1. Check that the `etcd` user and group exists on the host. Run `id etcd` to see if the etcd user and group exists.

> 2. Check etcd's data directory to see if it's owned by `etcd` and set to the correct permissions. Run `stat -c %U:%G /var/lib/rancher/k3s/server/db/etcd` to check the etcd data-dir owner and group, and run `stat -c %a /var/lib/rancher/k3s/server/db/etcd` to verify it has the correct permissions.

Reviewer 2: `ls` would show both owner and permissions, vs having to use two `stat` commands. Also, can we say what the correct permissions are?
```suggestion
2. Check etcd's data directory to see if it's owned by `etcd` and set to the correct permissions. Run `ls -la /var/lib/rancher/k3s/server/db/etcd` and verify the owner, group, and permissions.
```

3. Ensure the etcd process is ran as the `etcd` user and group by setting the etcd static pod's SecurityContext approopriately.

If the user and group don't exist, the command below can be run to create them.

```bash
useradd -r -c "etcd user" -s /sbin/nologin -M etcd
```

## Kubernetes Runtime Requirements

> The runtime requirements to comply with the CIS Benchmark are centered around pod security (PSP's)and network policies. These are outlined in this section. K3s doesn't apply any default PSP's or network policies however K3s ships with a controller that is meant to apply a given set of network policies. By default, K3s runs with the "NodeRestriction" admission controller. To enable PSP's, add the following to the K3s start command: `--kube-apiserver-arg="enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount"`. This will have the effect of maintaining the "NodeRestriction" plugin as well as enabling the "PodSecurityPolicy".

Reviewer 2: PSPs are plural, not possessive or a contraction, so it doesn't need an apostrophe.
```suggestion
The runtime requirements to comply with the CIS Benchmark are centered around pod security (PSPs) and network policies. These are outlined in this section. K3s doesn't apply any default PSPs or network policies, however K3s ships with a controller that is required to enforce any network policies provided by the user. By default, K3s runs with the "NodeRestriction" admission controller. To enable PSPs, add the following to the K3s start command: `--kube-apiserver-arg="enable-admission-plugins=NodeRestriction,PodSecurityPolicy,ServiceAccount"`. This will have the effect of maintaining the "NodeRestriction" plugin as well as enabling the "PodSecurityPolicy".
```

This can be remeditated by passing this argument as a value to the `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.
</details>

### Control 1.2.27
Ensure that the `--service-account-lookup` argument is set to true.
<details>
<summary>Rationale</summary>
If `--service-account-lookup` is not enabled, the apiserver only verifies that the authentication token is valid, and does not validate that the service account token mentioned in the request is actually present in etcd. This allows using a service account token even after the corresponding service account is deleted. This is an example of time of check to time of use security issue.

This can be remeditated by passing this argument as a value to the `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.
</details>

### Control 1.2.33

> Ensure that the `--encryption-provider-config` argument is set as appropriate.

Reviewer 2: Do we want to talk about the `--secrets-encryption` option? That will configure the encryption provider for the user.
https://rancher.com/docs/k3s/latest/en/advanced/#secrets-encryption-config-experimental

<details>
<summary>Rationale</summary>
Where `etcd` encryption is used, it is important to ensure that the appropriate set of encryption providers is used. Currently, the aescbc, kms and secretbox are likely to be appropriate options.
</details>

### Control 1.2.34
Ensure that encryption providers are appropriately configured.
<details>
<summary>Rationale</summary>
`etcd` is a highly available key-value store used by Kubernetes deployments for persistent storage of all of its REST API objects. These objects are sensitive in nature and should be encrypted at rest to avoid any disclosures.

> This can be remeditated by passing a valid configuration to the `k3s` outlined above.

Reviewer 2: ```suggestion
This can be remediated by passing a valid configuration to `k3s` as outlined above.
```

    --container-runtime-endpoint=/run/k3s/containerd/containerd.sock 
    --container-runtime=remote 
    --containerd=/run/k3s/containerd/containerd.sock 
    --eviction-hard=imagefs.available<5%,nodefs.available<5% 
    --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% 
    --fail-swap-on=false 
    --healthz-bind-address=127.0.0.1 
    --hostname-override=rke201 
    --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig 
    --kubelet-cgroups=/systemd/system.slice 
    --node-labels= 
    --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests 
    --protect-kernel-defaults=true                                                                   # 4.2.6
    --read-only-port=0                                                                               # 4.2.4
    --resolv-conf=/run/systemd/resolve/resolv.conf 
    --runtime-cgroups=/systemd/system.slice 
    --serialize-image-pulls=false 
    --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt                                   # 4.2.10
    --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key                            # 4.2.10
```

> An example command to be run that would providle the additional configuration and settings necessary to comply with CIS 1.5 can be referenced below. This code snippet is what would then be viewed on a running k3s system when issuing a `ps -ef | grep k3s` which would be used to validate that the current running state of the K3s node is CIS 1.5 compliant.

Reviewer 2: K3s hide its command-line flags at startup using setproctitle so we should not recommend checking `ps -ef` to see the flags on the running command. The best way is probably to check the annotation on the node:
```
kubectl get nodes -o 'template={{range .items}}{{.metadata.name}}: {{index .metadata.annotations "k3s.io/node-args"}}{{"\n"}}{{end}}'
```

Author: Crap. I forgot about the `setproctitle` call. That `kubectl` w/`jq` command though seems promising. gonna play with that a bit more. Thanks.

---
title: "CIS Hardening Guide"
weight: 80
---

> # CIS Hardening Guide

Reviewer 4: Remove lines 5 and 6 here. Due to the way this renders with hugo, we will see the title twice, as pictured below.

![image](https://user-images.githubusercontent.com/45179589/109063980-0f1b5e80-76a7-11eb-96ab-eb187a875af1.png)

Author: Updated.

This document provides prescriptive guidance for hardening a production installation of K3s. It outlines the configurations and controls required to address Kubernetes benchmark controls from the Center for Information Security (CIS).

K3s has a number of security mitigations applied and turned on by default and will pass a number of the Kubernetes CIS controls without modification. There are some notable exceptions to this that require manual intervention to fully comply with the CIS Benchmark:

1. K3s will not modify the host operating system. Any host-level modifications will need to be done manually.

> 2. Certain CIS policy controls for PodSecurityPolicies and NetworkPolicies will restrict the functionality of this cluster. You must opt into having K3s configure these by adding the appropriate options (enabliong of admission plugins) to your command-line flags or configuration file as well as manually applying appropriate policies. Further detail in the sections below.

Reviewer 4: Spelling

```suggestion
2. Certain CIS policy controls for PodSecurityPolicies and NetworkPolicies will restrict the functionality of this cluster. You must opt into having K3s configure these by adding the appropriate options (enabling of admission plugins) to your command-line flags or configuration file as well as manually applying appropriate policies. Further detail in the sections below.
```

Author: Updated.

> The first section (1.1) of the CIS Benchmark concerns itself primarily with pod manifest permissions and ownership. K3s doesn't utilize these for the core components since everything is packaged into a single binary.

Reviewer 4: Here and one or more times in the doc we mention the CIS Benchmark -- perhaps this is too nit picky, but should we be linking to it? When I read these mentions of sections of the CIS Benchmark, as a reader I don't necessarily know what that is or what is being referred to. I notice it mentioned in PodSecurityPolicies section as well (section 5.2)

Author: You need to sign up with a valid email and fill out a form requesting the Benchmark. It's not something you can link to.

Reviewer 4: Copy that. thanks! Resovling this.

## Host-level Requirements

> There are two areas of of host-level requirements: kernel parameters and etcd process/directory configuration. These are outlined in this section.

Reviewer 4: Duplicate word

```suggestion
There are two areas of host-level requirements: kernel parameters and etcd process/directory configuration. These are outlined in this section.
```

Author: Updated.

### Ensure `protect-kernel-defaults` is set

This is a kubelet flag that will cause the kubelet to exit if the required kernel parameters are unset or are set to values that are different from the kubelet's defaults.

> **Note:** `protect-kernel-defaults` is exposed as a top-level flag for K3s.

#### Set kernel parameters

> Create a file called `/etc/sysctl.d/90-kubelet.conf` and add the snippet below. Then run `sysctl -p /etc/sysctl.d/90-kubelet.conf` to 

Reviewer 4: Just want to double-check this sentence ends with "...to" was this expected? I suppose because you indicate what is being done in the bash code block below.

Author: Updated.

### PodSecurityPolicies

When PSPs are enabled, a policy can be applied to satisfy the necessary controls described in section 5.2 of the CIS Benchmark.

Here's an example of a compliant PSP.

```yaml
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: cis1.5-compliant-psp
spec:
  privileged: false                # CIS - 5.2.1
  allowPrivilegeEscalation: false  # CIS - 5.2.5
  requiredDropCapabilities:        # CIS - 5.2.7/8/9
    - ALL
  volumes:
    - 'configMap'
    - 'emptyDir'
    - 'projected'
    - 'secret'
    - 'downwardAPI'
    - 'persistentVolumeClaim'
  hostNetwork: false               # CIS - 5.2.4
  hostIPC: false                   # CIS - 5.2.3
  hostPID: false                   # CIS - 5.2.2
  runAsUser:
    rule: 'MustRunAsNonRoot'       # CIS - 5.2.6
  seLinux:
    rule: 'RunAsAny'
  supplementalGroups:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  fsGroup:
    rule: 'MustRunAs'
    ranges:
      - min: 1
        max: 65535
  readOnlyRootFilesystem: false
```

> Before the above PSP to be effective, we need to create a couple ClusterRoles and ClusterRole. We also need to include a "system unrestricted policy" which is needed for system level pods that require additional privileges.

Reviewer 4: Nit - hyphen usage

```suggestion
Before the above PSP to be effective, we need to create a couple ClusterRoles and ClusterRole. We also need to include a "system unrestricted policy" which is needed for system-level pods that require additional privileges.
```

Author: Updated.

  volumes:
  - '*'
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: system-unrestricted-node-psp-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system-unrestricted-psp-role
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:nodes
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: system-unrestricted-psp-role
rules:
- apiGroups:
  - policy
  resourceNames:
  - system-unrestricted-psp
  resources:
  - podsecuritypolicies
  verbs:
  - use
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: system-unrestricted-svc-acct-psp-rolebinding
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system-unrestricted-psp-role
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:serviceaccounts
```

> **Note:** The Kubernetes critical additions such as CNI, DNS, and Ingress are ran as pods in the `kube-system` namespace. Therefore, this namespace will have a policy that is less restrictive so that these components can run properly.

### NetworkPolicies

> > NOTE: K3s's deploys kube-router for network policy enforcement. Support for this in K3s is currently experimental.

Reviewer 4: Possible mistake with the utilization of an apostrophie

```suggestion
> NOTE: K3s deploys kube-router for network policy enforcement. Support for this in K3s is currently experimental.
```

Author: Updated.

Here's an example of a compliant network policy. 

```yaml
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: intra-namespace
  namespace: kube-system
spec:
  podSelector: {}
  ingress:
    - from:
      - namespaceSelector:
          matchLabels:
            name: kube-system
```

> **Note:** Operators must manage network policies as normal for additional namespaces that are created.

## Known Issues
The following are controls that K3s currently does not pass by default. Each gap will be explained, along with a note clarifying whether it can be passed through manual operator intervention, or if it will be addressed in a future release of K3s.

### Control 1.2.15
Ensure that the admission control plugin `NamespaceLifecycle` is set.
<details>
<summary>Rationale</summary>
Setting admission control policy to NamespaceLifecycle ensures that objects cannot be created in non-existent namespaces, and that namespaces undergoing termination are not used for creating the new objects. This is recommended to enforce the integrity of the namespace termination process and also for the availability of the newer objects.

> This can be remeditated by passing this argument as a value to the `enable-admission-plugins=` and pass that to  `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.

Reviewer 4: Spelling

```suggestion
This can be remediated by passing this argument as a value to the `enable-admission-plugins=` and pass that to  `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.
```

Reviewer 4: This is also seen throughout the document. Search for remediated and fix throughout please. Probably was a copy and paste situation :)

Reviewer 4: Nit: I see many sentences in the Known Issues section stating "An example can be found below." Maybe it's just me, but I expect generally to see an example right below. I believe you are referring to examples outlined in the Control Plane Execution and Argument section. If so -- do you think we should refer to that section?

I'm calling this a nit, but as I read the document it's not very apparent to me to where I need to scroll down to to see my example.

Author: Updated.

`etcd` is a highly available key-value store used by Kubernetes deployments for persistent storage of all of its REST API objects. These objects are sensitive in nature and should be encrypted at rest to avoid any disclosures.

This can be remediated by passing a valid configuration to `k3s` as outlined above.
</details>

### Control 1.3.1
Ensure that the `--terminated-pod-gc-threshold` argument is set as appropriate.
<details>
<summary>Rationale</summary>
Garbage collection is important to ensure sufficient resource availability and avoiding degraded performance and availability. In the worst case, the system might crash or just be unusable for a long period of time. The current setting for garbage collection is 12,500 terminated pods which might be too high for your system to sustain. Based on your system resources and tests, choose an appropriate threshold value to activate garbage collection.

This can be remeditated by passing this argument as a value to the `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.
</details>

### Control 3.2.1
Ensure that a minimal audit policy is created (Scored)
<details>
<summary>Rationale</summary>

> Logging is an important detective control for all systems, to detect potential unauthorised access.

Reviewer 4: Spelling

```suggestion
Logging is an important detective control for all systems, to detect potential unauthorized access.
```

Author: Likely from the PDF and the PDF was written by a Brit. :)

This can be remeditated by passing this argument as a value to the `--kube-apiserver-arg=` argument to `k3s server`. An example can be found below.
</details>

### Control 5.1.5
Ensure that default service accounts are not actively used. (Scored)
<details>
<summary>Rationale</summary>

Kubernetes provides a default service account which is used by cluster workloads where no specific service account is assigned to the pod.

Where access to the Kubernetes API from a pod is required, a specific service account should be created for that pod, and rights granted to that service account.

The default service account should be configured such that it does not provide a service account token and does not have any explicit rights assignments.
</details>

The remediation for this is to update the `automountServiceAccountToken` field to `false` for the `default` service account in each namespace.

> For `default` service accounts in the built-in namespaces (`kube-system`, `kube-public`, `kube-node-lease`, and `default`), K3s does not automatically do this. You can manually update this field on these service accounts to passs the control.

Reviewer 4: Spelling

```suggestion
For `default` service accounts in the built-in namespaces (`kube-system`, `kube-public`, `kube-node-lease`, and `default`), K3s does not automatically do this. You can manually update this field on these service accounts to pass the control.
```

Author: Updated.

## Control Plane Execution and Arguments

Listed below are the K3s control plane components and the arguments they're given at start, by default. Commented to their right is the CIS 1.5 control that they satisfy.

> ```bash

Reviewer 4: Nit: When rendered in the browser the comments are far to the right causing a need for horizontal scrolling... This isn't really fair to ask but maybe we can make the comments closer to the left? If not no worries. It just took me a while to figure out that the issue was the bad CSS on the rancher.com website and so I had to hunt for my tiny horizontal scroll bar to see these comments.

![image](https://user-images.githubusercontent.com/45179589/109065614-180d2f80-76a9-11eb-8e55-6b3efe53f586.png)

![image](https://user-images.githubusercontent.com/45179589/109065647-20656a80-76a9-11eb-93cb-cf8463261140.png)

Author: I brought the comments in as far as possible to keep them in a column and lined up.

    --serialize-image-pulls=false 
    --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt                                   # 4.2.10
    --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key                            # 4.2.10
```

The command below is an example of how the outlined remediations can be applied.

```bash
k3s server \
    --protect-kernel-defaults=true \
    --secrets-encryption=true \
    --kube-apiserver-arg='audit-log-path=/var/lib/rancher/k3s/server/logs/audit-log' \
    --kube-apiserver-arg='audit-log-maxage=30' \
    --kube-apiserver-arg='audit-log-maxbackup=10' \
    --kube-apiserver-arg='audit-log-maxsize=100' \
    --kube-apiserver-arg='request-timeout=300s' \
    --kube-apiserver-arg='service-account-lookup=true' \
    --kube-apiserver-arg='enable-admission-plugins=NodeRestriction,PodSecurityPolicy,NamespaceLifecycle,ServiceAccount' \
    --kube-controller-manager-arg='terminated-pod-gc-threshold=10' \
    --kube-controller-manager-arg='use-service-account-credentials=true' \
    --kubelet-arg='streaming-connection-idle-timeout=5m' \
    --kubelet-arg='make-iptables-util-chains=true'
```

## Conclusion

> If you have followed this guide, your K3s cluster will be configured to comply with the CIS Kubernetes Benchmark. You can review the [CIS Benchmark Self-Assessment Guide](../self_assessment/_index.md) to understand the expectations of each of the benchmarks and how you can do the same on your cluster.

Reviewer 4: Bad link

```suggestion
If you have followed this guide, your K3s cluster will be configured to comply with the CIS Kubernetes Benchmark. You can review the [CIS Benchmark Self-Assessment Guide](../self_assessment/) to understand the expectations of each of the benchmarks and how you can do the same on your cluster.
```

Author: Updated.
